workspaces/libnpmpublish/lib/provenance.js

```diff
@@ -1,68 +1,208 @@
+        builder: { id: GITHUB_BUILDER_ID },
+        invocation: {
+          configSource: {
+            uri: `git+${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}@${env.GITHUB_REF}`,
+            digest: {
+              sha1: env.GITHUB_SHA,
+            },
+            entryPoint: workflowPath,
+          },
+          parameters: {},
+          environment: {
+            GITHUB_EVENT_NAME: env.GITHUB_EVENT_NAME,
+            GITHUB_REF: env.GITHUB_REF,
+            GITHUB_REPOSITORY: env.GITHUB_REPOSITORY,
+            GITHUB_REPOSITORY_ID: env.GITHUB_REPOSITORY_ID,
+            GITHUB_REPOSITORY_OWNER_ID: env.GITHUB_REPOSITORY_OWNER_ID,
+            GITHUB_RUN_ATTEMPT: env.GITHUB_RUN_ATTEMPT,
+            GITHUB_RUN_ID: env.GITHUB_RUN_ID,
+            GITHUB_SHA: env.GITHUB_SHA,
+            GITHUB_WORKFLOW_REF: env.GITHUB_WORKFLOW_REF,
+            GITHUB_WORKFLOW_SHA: env.GITHUB_WORKFLOW_SHA,
           },
-          entryPoint: workflowPath,
         },
-        parameters: {},
-        environment: {
-          GITHUB_EVENT_NAME: env.GITHUB_EVENT_NAME,
-          GITHUB_REF: env.GITHUB_REF,
-          GITHUB_REPOSITORY: env.GITHUB_REPOSITORY,
-          GITHUB_REPOSITORY_ID: env.GITHUB_REPOSITORY_ID,
-          GITHUB_REPOSITORY_OWNER_ID: env.GITHUB_REPOSITORY_OWNER_ID,
-          GITHUB_RUN_ATTEMPT: env.GITHUB_RUN_ATTEMPT,
-          GITHUB_RUN_ID: env.GITHUB_RUN_ID,
-          GITHUB_SHA: env.GITHUB_SHA,
-          GITHUB_WORKFLOW_REF: env.GITHUB_WORKFLOW_REF,
-          GITHUB_WORKFLOW_SHA: env.GITHUB_WORKFLOW_SHA,
+        metadata: {
+          buildInvocationId: `${env.GITHUB_RUN_ID}-${env.GITHUB_RUN_ATTEMPT}`,
+          completeness: {
+            parameters: false,
+            environment: false,
+            materials: false,
+          },
+          reproducible: false,
         },
+        materials: [
+          {
+            uri: `git+${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}@${env.GITHUB_REF}`,
+            digest: {
+              sha1: env.GITHUB_SHA,
+            },
+          },
+        ],
       },
-      metadata: {
-        buildInvocationId: `${env.GITHUB_RUN_ID}-${env.GITHUB_RUN_ATTEMPT}`,
-        completeness: {
-          parameters: false,
-          environment: false,
-          materials: false,
+    }
+  } else if (ci.GITLAB) {
+    // TODO: pull values from authenticated JWT rather than environment variables.
+    payload = {
+      _type: INTOTO_STATEMENT_TYPE,
+      subject,
+      predicateType: SLSA_PREDICATE_TYPE,
+      predicate: {
+        buildType: `${GITLAB_BUILD_TYPE_PREFIX}/${GITLAB_BUILD_TYPE_VERSION}`,
+        builder: { id: `${env.CI_PROJECT_URL}/-/runners/${env.CI_RUNNER_ID}` },
+        invocation: {
+          configSource: {
+            uri: `git+${env.CI_PROJECT_URL}@${env.CI_BUILD_REF}`,
+            digest: {
+              sha1: env.CI_COMMIT_SHA,
+            },
+            entryPoint: env.CI_JOB_NAME,
+          },
+          parameters: {
```

Comment:
According to the [slsa 0.2 spec](https://slsa.dev/provenance/v0.2), `parameters` are external inputs that influenced the build on top of `invocation.configSource`.

Wondering if most of these should be moved to `environment` instead, as it's not clear how exactly they might influence the build.

Also, there's a lot here! Do we really need to include all these values? Could we prune this down to things that are crucial for provenance, i.e. understanding where and how it was produced.

Could you provide a complete example that might make it easier to spot what's useful?

cc @ianlewis @MarkLodato @laurentsimon would love your input here as well. Don't know how much time you've spent looking at how a provenance statement should look for gitlab? Maybe we should draft up a 1.0 spec for gitlab?

Reply:
GitLab does some auto-scrubbing to prevent secrets set in environment vars from leaking out.

These are explicitly listed as a simplified way to emulate that behavior so we're not accidentally including secret values.

Reply:
I agree these would generally be the "inputs" to a workflow. Everything `CI*` and `GITLAB_*` would be better placed in the `environment` imho.

This is much clearer in SLSA v1.0 where it's clearly delineated between parameters under user control (`externalParameters`) and those under the platform's control (`internalParameters`). In 0.2 it's a bit more fuzzy but `environment` generally maps to the `internalParameters` and `parameters` maps to `externalParameters`.

Reply:
> GitLab SLSA v0.2 spec

Ah gotcha, keeping parity with the existing gitlab v0.2 schema makes sense to me unless we want to go to 1.0 but sounds like this might take some time to settle on 👍